guide/adopting-the-configuration-cache.md

```diff
@@ -0,0 +1,384 @@
+      // Do something with `result`...
+      Path projectDir = getProjectLayout().getProjectDirectory().getAsFile().toPath();
+      // Do something with `projectDir`...
+   }
+}
+```
+
+
+## Two strategies to write Configuration Cache friendly Gradle
+
+To summarize the two strategies we used to solve Configuration Cache issues:
+
+### You can inject services into Gradle-managed types
+
+Do you need the project directory? Inject [`ProjectLayout`](https://docs.gradle.org/current/javadoc/org/gradle/api/file/ProjectLayout.html). Do you need to run a bash command? Inject [`ExecOperations`](https://docs.gradle.org/current/javadoc/org/gradle/process/ExecOperations.html). Gradle offers a [list](https://docs.gradle.org/current/userguide/service_injection.html) of injectable services which will cover 99% of your use cases. However, that list is incomplete — in reality,  almost [anything in Gradle source](https://github.com/search?q=repo%3Agradle%2Fgradle%20%40ServiceScope&type=code) annotated with [`@ServiceScope`](https://github.com/gradle/gradle/blob/196bb409d47f5b6e39d62edd39be939f7606a5cc/platforms/core-runtime/stdlib-java-extensions/src/main/java/org/gradle/internal/service/scopes/ServiceScope.java#L43) can be injected into a Gradle managed type.
+
+
+### You can make any class into a Gradle-managed type
+
+A standard class...
+
+```java
+class Employee {
+    private String name;
+
+    public Employee(String name) {
+        this.name = name;
+    }
+}
+```
+
+...instantiated like so...
+
+```java
+Employee employee = new Employee("Kelvin");
+```
+
+...can be turned into a Gradle-managed type by making it abstract, and adding `@Inject` to the constructor.
+
+```java
+abstract class Employee {
+    private String name;
+
+    @Inject
+    public Employee(String name) {
+        this.name = name;
+    }
+}
+```
+
+Now, we can instantiate it with `ObjectFactory`.
+
+```java
+Employee employee = objectFactory.newInstance(Employee.class, "Kelvin");
+```
+
+Hold on, what if the class using `Employee` doesn't have access to an `ObjectFactory`? 
+
+```java
+class Team {
+    private ArrayList<Employee> employees = new ArrayList<>();
+
+    public addEmployee() {
+        Employee employee = new Employee("Kelvin");
+        employees.add(employee);
+    }
+}
+```
+
+In that case, you can turn *it* a Gradle managed type as well...
+
+```java
+abstract class Team {
+    private ArrayList<Employee> employees = new ArrayList<>();
+
+    @Inject
+    protected abstract ObjectFactory getObjectFactory();
+
+    public addEmployee() {
+        Employee employee = getObjectFactory().newInstance(Employee.class, "Kelvin");
+        employees.add(employee);
+    }
+}
+```
+
+...so on and so forth...
+
+```java
+abstract class Deployment {
+    private String customerName;
+
+    @Nested
+    public abstract Team getTeam();
+ 
+    @Inject
+    public Deployment(String customerName) {
+        this.customerName = customerName; 
+    }
+}
+
+```
+
+> [!TIP]
```

Comment:
I think you're underselling this!

This probably deserves its own section and/or have most of the examples work via `@Nested`. This is the end goal of where you will probably want to end up in most cases.